docs/source/reference/ha.rst

```diff
@@ -0,0 +1,301 @@
+^^^^^^^^^^^^^^^
+Various |st2| features rely on a proper co-ordination backend in a distributed deployment to work correctly.
+
+`This <http://zookeeper.apache.org/doc/trunk/zookeeperStarted.html#sc_RunningReplicatedZooKeeper>`__ shows how to run a replicated zookeeper setup. See `this <http://redis.io/topics/sentinel>`__ to understand Redis deployments using sentinel.
+
+
+Nginx and loadbalancer
+^^^^^^^^^^^^^^^^^^^^^^^
+An Nginx server is required to reverse proxy each instance of ``st2api``, ``st2auth``, ``st2stream`` and ``mistral-api``. This server will terminate SSL connections, shield clients from internal port numbers of various services and only require ports 80 and 443 to be open on containers. Often it is best to deploy 1 set of all these services on a compute instance and share an Nginx server.
+
+There is also a need for a loadbalancer to frontend all the REST services. This results in an HA deployment for REST services as well as single endpoint for clients. Most deployment infrastructures will already have a loadbalancer solution which they would
+prefer to use so we do not provide any recommendations.
+
+
+Reference HA setup
+------------------
+
+In this section we provide a highly opinionated and therefore prescriptive approach to deploying |st2| in HA. This deployment
+has 3 independent boxes which we categorize as controller box and blueprint box. Let call these boxes ``st2-multi-node-cntl``,
+``st2-multi-node-1`` and ``st2-multi-node-2``. For the sake of reference we will be using Ubuntu 14.04 as the base operating
+system.
+
+.. figure :: /_static/images/st2-deployment-multi-node.png
+    :align: center
+
+    StackStorm HA reference deployment.
+
+.. source https://docs.google.com/drawings/d/1_BJa9ZtBjFa1Dxx6cPiFlmpTS9AsNzkkvp_vuyVV3bw/edit
+
+Controller box
+^^^^^^^^^^^^^^
+This box runs all the shared required dependencies and some |st2| components.
+
+* Nginx as loadbalancer
+* MongoDB
+* PostgreSQL
+* RabbitMQ
+* st2chatops
+* st2web
+
+In practice ``MongoDB``, ``PostgreSQL`` and ``RabbitMQ`` are often in standalone clusters managed outside of |st2|.
+The 2 shared components i.e. ``st2chatops`` and ``st2web`` are placed here for sake of convenience and could be placed anywhere
+with the right configuration.
+
+Nginx acting as the loadbalancer can easily be switched out for Amazon ELB, HAProxy or any other of your choosing. In that case ``st2web`` which is being served off this Nginx will also need a new home.
+
+``st2chatops`` which use ``hubot`` is not easily deployed in HA. Using something like `keepalived <http://www.keepalived.org/>`__
+to maintain st2chatops in active-passive configuration would be an option.
+
+Following are the steps to provision a controller box on Ubuntu 14.04.
+
+Install required dependencies
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+1. Install ``MongoDB``, ``PostgreSQL`` and ``RabbitMQ``.
+
+  .. code-block:: bash
+
+        sudo apt-get install -y mongodb-server rabbitmq-server postgresql
```

Comment:
I'm personally not a fan of having any of those "low level" instructions in this document since it's way too easy for those things to get out of sync with the reality.